iceoryx_dds/source/iceoryx_dds/dds/cyclone_data_reader.cpp

```diff
@@ -13,7 +13,111 @@
+        auto topic = ::dds::topic::Topic<Mempool::Chunk>(CycloneContext::getParticipant(), topicString);
+        auto subscriber = ::dds::sub::Subscriber(CycloneContext::getParticipant());
+
+        auto qos = ::dds::sub::qos::DataReaderQos();
+        qos << ::dds::core::policy::History::KeepAll();
+
+        m_impl = ::dds::sub::DataReader<Mempool::Chunk>(subscriber, topic, qos);
+
+        LogDebug() << "[CycloneDataReader] Connected to topic: " << topicString;
+
+        m_isConnected.store(true, std::memory_order_relaxed);
+    }
+}
+
+iox::cxx::expected<uint64_t, iox::dds::DataReaderError>
+iox::dds::CycloneDataReader::read(uint8_t* const buffer, const uint64_t& bufferSize, const uint64_t& sampleSize)
+{
+    auto bufferCapacity = bufferSize / sampleSize;
+    return read(buffer, bufferSize, sampleSize, bufferCapacity);
+}
+
+iox::cxx::expected<uint64_t, iox::dds::DataReaderError> iox::dds::CycloneDataReader::read(uint8_t* const buffer,
+                                                                                          const uint64_t& bufferSize,
+                                                                                          const uint64_t& sampleSize,
+                                                                                          const uint64_t& maxSamples)
+{
+    // Validation checks
+    if (!m_isConnected.load())
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::NOT_CONNECTED);
+    }
+    if (buffer == nullptr)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+    if (bufferSize < sampleSize)
+    {
+        return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_RECV_BUFFER);
+    }
+
+    // Read up to the maximum number of samples that can fit in the buffer.
+    auto samples = m_impl.select()
+                       .max_samples(static_cast<uint32_t>(maxSamples))
+                       .state(::dds::sub::status::SampleState::not_read())
+                       .take();
+
+    // Copy data into the provided buffer.
+    uint64_t numSamplesBuffered = 0u;
+    if (samples.length() > 0)
+    {
+        // Sample validation checks
+        uint64_t size = samples.begin()->data().payload().size();
+        if (size != sampleSize)
+        {
+            // Received invalid data.
+            // NOTE: This causes other data points received in this read to be lost...
+            return iox::cxx::error<iox::dds::DataReaderError>(iox::dds::DataReaderError::INVALID_DATA);
+        }
+
+        // Do copy
+        uint64_t cursor = 0; // Tracks the position in the buffer to write next sample.
+        for (const auto& sample : samples)
+        {
+            auto bytes = sample.data().payload().data();
+            std::copy(bytes, bytes + sampleSize, &buffer[cursor]);
```

Comment:
So you are doing the copy here. Could you just take a sample from the reader, then check the payload size, then allocate a chunk with this size and copy the dynamic payload size?